### ISSUE TITLE
Docker daemon hangs. Possible deadlock in 1.11.2

### ISSUE THREAD

**Issue by sagar8192** [moby/moby#27566]
**Description:**
Docker daemon becomes unresponsive on boot `and` all the requests hang forever.

**Ways to reproduce:**
I have not been able to reproduce it yet.

**Describe the results you received:**
Docker daemon becomes unresponsive/hangs.

Cannot run docker info or docker version on the host. But, this is the output from another host.

```
Docker version:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   e2d06fe
 Built:        Tue Jun 14 17:49:14 2016
 OS/Arch:      linux/amd64
```

```
Docker info:
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 51
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /ephemeral/docker/aufs
 Backing Filesystem: xfs
 Dirs: 237
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-34-generic
Operating System: <unknown> (containerized)
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 120.1 GiB
ID: 6LCF:UXLX:4ROM:UIYI:BRJK:MVTY:QKGA:PPIT:MVG4:ZQTC:6XHS:6LVG
Docker Root Dir: /ephemeral/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
```

```
SIGQUIT: quit
 PC=0x46fbe1 m=0

 goroutine 0 [idle]:
 runtime.futex(0x2296f90, 0x0, 0x0, 0x0, 0x0, 0x2296380, 0x0, 0x0, 0x41d9d4, 0x2296f90, ...)
  /usr/local/go/src/runtime/sys_linux_amd64.s:288 +0x21
 runtime.futexsleep(0x2296f90, 0x0, 0xffffffffffffffff)
  /usr/local/go/src/runtime/os1_linux.go:39 +0x53
 runtime.notesleep(0x2296f90)
  /usr/local/go/src/runtime/lock_futex.go:142 +0xa4
 runtime.stopm()
  /usr/local/go/src/runtime/proc1.go:1136 +0x112
 runtime.findrunnable(0xc820017500, 0x0)
  /usr/local/go/src/runtime/proc1.go:1538 +0x69e
 runtime.schedule()
  /usr/local/go/src/runtime/proc1.go:1647 +0x267
 runtime.park_m(0xc820000180)
  /usr/local/go/src/runtime/proc1.go:1706 +0x18b
 runtime.mcall(0x7ffea96f3240)
  /usr/local/go/src/runtime/asm_amd64.s:204 +0x5b

 goroutine 1 [select]:
 google.golang.org/grpc.(*Conn).Wait(0xc8203dc0c0, 0x7fa664356200, 0xc820011000, 0x0, 0x0, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:530 +0x34b
 google.golang.org/grpc.(*unicastPicker).Pick(0xc8204336a0, 0x7fa664356200, 0xc820011000, 0x0, 0x0, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/picker.go:84 +0x51
 google.golang.org/grpc.NewClientStream(0x7fa664356200, 0xc820011000, 0x2251fe0, 0xc82035b400, 0x18a2660, 0x11, 0x0, 0x0, 0x0, 0x0, ...)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/stream.go:103 +0xb8
 github.com/docker/containerd/api/grpc/types.(*aPIClient).Events(0xc8203561a0, 0x7fa664356200, 0xc820011000, 0xc820444e90, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  /go/src/github.com/docker/docker/vendor/src/github.com/docker/containerd/api/grpc/types/api.pb.go:930 +0xe9
 github.com/docker/docker/libcontainerd.(*remote).startEventsMonitor(0xc82036c4d0, 0x0, 0x0)
  /go/src/github.com/docker/docker/libcontainerd/remote_linux.go:247 +0x106
 github.com/docker/docker/libcontainerd.New(0xc820433460, 0x1d, 0xc820433480, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0)
  /go/src/github.com/docker/docker/libcontainerd/remote_linux.go:105 +0xa32
 main.(*DaemonCli).CmdDaemon(0xc820429970, 0xc82000a290, 0xb, 0xb, 0x0, 0x0)
  /go/src/github.com/docker/docker/docker/daemon.go:269 +0x19ec
 reflect.callMethod(0xc820417860, 0xc8201ffc28)
  /usr/local/go/src/reflect/value.go:628 +0x1fc
 reflect.methodValueCall(0xc82000a290, 0xb, 0xb, 0x1, 0xc820417860, 0x0, 0x0, 0xc820417860, 0x0, 0x4758c4, ...)
  /usr/local/go/src/reflect/asm_amd64.s:29 +0x36
 github.com/docker/docker/cli.(*Cli).Run(0xc8204177d0, 0xc82000a280, 0xc, 0xc, 0x0, 0x0)
  /go/src/github.com/docker/docker/cli/cli.go:89 +0x383
 main.main()
  /go/src/github.com/docker/docker/docker/docker.go:63 +0x43c

 goroutine 17 [syscall, locked to thread]:
 runtime.goexit()
  /usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

 goroutine 19 [sleep]:
 time.Sleep(0x20030225b4)
  /usr/local/go/src/runtime/time.go:59 +0xf9
 google.golang.org/grpc.(*Conn).resetTransport(0xc8203dc0c0, 0x17b8100, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:440 +0x77f
 google.golang.org/grpc.NewConn.func1(0xc8203dc0c0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:296 +0x2a
 created by google.golang.org/grpc.NewConn
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:302 +0x4fb

 goroutine 6 [syscall]:
 os/signal.loop()
  /usr/local/go/src/os/signal/signal_unix.go:22 +0x18
 created by os/signal.init.1
  /usr/local/go/src/os/signal/signal_unix.go:28 +0x37

 goroutine 20 [semacquire]:
 sync.runtime_Syncsemacquire(0xc820449dd0)
  /usr/local/go/src/runtime/sema.go:237 +0x201
 sync.(*Cond).Wait(0xc820449dc0)
  /usr/local/go/src/sync/cond.go:62 +0x9b
 google.golang.org/grpc.(*Conn).WaitForStateChange(0xc8203dc0c0, 0x7fa664356200, 0xc820011000, 0x3, 0x0, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:351 +0x19f
 google.golang.org/grpc.(*unicastPicker).WaitForStateChange(0xc8204336a0, 0x7fa664356200, 0xc820011000, 0x3, 0x48688a, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/picker.go:96 +0x4f
 google.golang.org/grpc.(*ClientConn).WaitForStateChange(0xc82035b400, 0x7fa664356200, 0xc820011000, 0x3, 0x1, 0x0, 0x0)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:228 +0x6b
 github.com/docker/docker/libcontainerd.(*remote).handleConnectionChange(0xc82036c4d0)
  /go/src/github.com/docker/docker/libcontainerd/remote_linux.go:116 +0xa5
 created by github.com/docker/docker/libcontainerd.New
  /go/src/github.com/docker/docker/libcontainerd/remote_linux.go:103 +0xa24

 goroutine 8 [select]:
 google.golang.org/grpc.(*Conn).WaitForStateChange.func1(0x7fa664356200, 0xc820011000, 0xc8203dc0c0, 0xc82033e5d0, 0xc820211b00)
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:340 +0x1b0
 created by google.golang.org/grpc.(*Conn).WaitForStateChange
  /go/src/github.com/docker/docker/vendor/src/google.golang.org/grpc/clientconn.go:348 +0x14e

<End of stack dump>
```

From the stack dump it looks like goroutine 20 and goroutine 8 are stuck waiting for WaitForStateChange. And goroutine 20 is holding the lock.

Strace from a client command:

```
Process 23781 attached - interrupt to quit
read(3, ^C <unfinished ...>

lr-x------ 1 root root 64 2016-10-20 08:53 /proc/23781/fd/3 -> pipe:[243784]

`lsof | grep 243784`
docker-do  23781          root    3r     FIFO               0,10      0t0     243784 pipe
docker     23784          root    1w     FIFO               0,10      0t0     243784 pipe
```

Docker daemon strace is not very helpful
`strace -ff -p`

```
Process 20195 attached with 7 threads - interrupt to quit
[pid 20220] restart_syscall(<... resuming interrupted call ...> <unfinished ...>
[pid 20205] futex(0xc820364110, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 20200] epoll_wait(4,  <unfinished ...>
[pid 20199] futex(0x22c8920, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 20198] futex(0xc820038a10, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 20197] restart_syscall(<... resuming interrupted call ...> <unfinished ...>
[pid 20195] futex(0x2296f90, FUTEX_WAIT, 0, NULL
```

Last lines before docker daemon gets stuck:

```
level=info msg="New containerd process, pid: 22473\n"
level=info msg="previous instance of containerd still alive (22473)"
```

Logs after docker daemon resumes its normal operation:

```
level=info msg="New containerd process, pid: 114140\n"
level=info msg="Graph migration to content-addressability took 0.00 seconds"
level=warning msg="mountpoint for pids not found"
level=info msg="Loading containers: start."
level=info msg="Loading containers: done."
level=info msg="Daemon has completed initialization"
```

Note: pid 22473 is not present on the box.

**Comment by sagar8192** [comment]
@justincormack Can someone please look into this? This is still happening for us.

**Comment by cpuguy83** [comment]
@sagar8192 It would be really helpful if you can try with 1.12.
There's been a number of fixes for issues with the containerd integration (which is what these seem to be hanging on).

**Comment by traffetseder** [comment]
We are running into the same issue while running in ECS (same docker logs and same stack dump). We are using Amazon ECS-optimized AMI and are using docker 1.11.2.

It's hard to reproduce and I can't tell what drives this to happen. Sometimes adding instances are hanging for 50+min and then recover, other times an instance boots and docker is available right away (with the exact same configuration).

**Comment by thaJeztah** [comment]
@traffetseder could you try as well if it's resolved in the current (1.12) release?

**Comment by thaJeztah** [comment]
closing; this one went stale
